blog/2023-07-31-new-pre-post-command-plugin-hooks.mdx

```diff
@@ -0,0 +1,165 @@
+    """Checks to see if the current environment is guarded"""
+    current_environment = get_current_environment()
+    
+    if is_guarded(current_environment):
+        raise CondaError(
+            f"Current environment is guarded. Run `conda guard {environment}` to "
+            "remove guard."
+        )
+
+@hookimpl
+def conda_pre_commands():
+    yield CondaPreCommand(
+        name=f"conda_guard_pre_command",
+        action=conda_guard_pre_commands_action,
+        run_for={"install", "remove", "update", "env_update", "env_remove"},
+    )
+```
+
+In the example above, we first register our plugin hook by defining a function called
+`conda_pre_commands` and then decorating it with the `hookimpl` function provided by conda.
+Inside the hook function itself we return a `CondaPreCommand` object, which defines the following:
+
+- **name**: how this plugin hook will be internally reference by conda
+- **action**: a callable object that will be invoked before the commands defined in `run_for` are
+  executed
+- **run_for**: a `set` of commands that this plugin hook should be used with
+
+Going back to our example, if we wanted to create a plugin that would prevent us from mistakenly
+modifying "guarded" environments, it would make sense to define `run_for` as all the conda commands
+that could potentially modify an environment (e.g. `install`, `remove`, etc.).
+
+To actually prevent us from modifying our guarded environments, the `action` function
+checks if the current environment is guarded and then throws a `CondaError` if it is to make
+sure that the program exits early with a meaningful error message for the user.
+
+:::info
+
+For the full implementation, please see the [conda-guard][conda-guard] project.
+
+:::
+
+## Simple analytics with "post command"
+
+The "post command" hook functions in exactly the same way as the "pre command" except that it gets
+called once a conda command has successfully finished running. It is important to note that if
+the command exits early for any reason, this plugin hook will not be invoked.
+
+To illustrate how this is used, we will create a simple plugin that counts how many times we have
+run a particular conda command. We can use this later to analyze our usage of conda 🤓 📊
+
+Here is a snippet of how that plugin might look:
+
+```python
+from conda.cli.conda_argparse import BUILTIN_COMMANDS
+from conda.plugins import hookimpl, CondaPostCommand
+
+ENV_COMMANDS = {
+    "env_config", "env_create", 
+    "env_export", "env_list",
+    "env_remove", "env_update"
+}
+
+def conda_stats_post_commands_action(command: str):
+    """Counts how many times we have run a particular conda command"""
+    database = get_database()
+    database.add_command(command)
+
+@hookimpl
+def conda_post_commands():
+    yield CondaPostCommand(
+        name=f"conda_stats_post_command",
+        action=conda_stats_post_commands_action,
+        run_for=BUILTIN_COMMANDS.union(ENV_COMMANDS),
+    )
+```
```

Comment:
```suggestion
@plugins.hookimpl
def conda_post_commands():
    yield plugins.CondaPostCommand(
        name=f"conda_stats_post_command",
        action=conda_stats_post_commands_action,
        run_for=BUILTIN_COMMANDS.union(ENV_COMMANDS),
    )
```

I think this makes it a little more readable since `hookimpl` is rather obtuse